src/NuGet.Core/NuGet.Build.Tasks/GetRestoreSolutionProjectsTask.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace NuGet.Build.Tasks
+{
+    public class GetRestoreSolutionProjectsTask : Task
+    {
+        private const string MetaProjExtension = ".metaproj";
+
+        /// <summary>
+        /// Solution project references.
+        /// </summary>
+        [Required]
+        public ITaskItem[] ProjectReferences { get; set; }
+
+        /// <summary>
+        /// Root project path used for resolving the absolute path.
+        /// </summary>
+        [Required]
+        public string SolutionFilePath { get; set; }
+
+        /// <summary>
+        /// Output items
+        /// </summary>
+        [Output]
+        public ITaskItem[] OutputProjectReferences { get; set; }
+
+        public override bool Execute()
+        {
+            // Log inputs
+            var log = new MSBuildLogger(Log);
+            log.LogDebug($"(in) ProjectReferences '{string.Join(";", ProjectReferences.Select(p => p.ItemSpec))}'");
+            log.LogDebug($"(in) SolutionFilePath '{SolutionFilePath}'");
+
+            var entries = new List<ITaskItem>();
+
+            // Filter obvious duplicates without considering OS case sensitivity.
+            // This will be filtered further when creating the spec.
+            var seen = new HashSet<string>(StringComparer.Ordinal);
+            var parentDirectory = Path.GetDirectoryName(SolutionFilePath);
+
+            foreach (var project in ProjectReferences)
+            {
+                if (string.IsNullOrEmpty(project.ItemSpec))
+                {
+                    continue;
+                }
+
+                var projectPath = Path.Combine(parentDirectory, project.ItemSpec);
+
+                // Check for the metaproj extension, this is auto generated by solutions instead of 
+                // the actual project path when build order is set. For the purpose of restore 
+                // the order is not important so we remove the extension to get the real project path.
+                if (projectPath.EndsWith(MetaProjExtension, StringComparison.OrdinalIgnoreCase))
+                {
+                    // Remove .metaproj from the path.
+                    projectPath = projectPath.Substring(0, projectPath.Length - MetaProjExtension.Length);
```

Comment:
i'd prefer being more verbose and call the helpers but your call.

---
In reply to: [107747427](https://github.com/NuGet/NuGet.Client/pull/1259#discussion_r107747427) [](ancestors = 107747427)